src/burnchains/bitcoin/address.rs

```diff
@@ -145,120 +222,406 @@ impl BitcoinAddress {
+        let mut version_bytes = Vec::with_capacity(1 + bytes.len());
+        version_bytes.push(version);
+        version_bytes.append(&mut bytes);
+        version_bytes
+    }
+
+    pub fn is_mainnet(&self) -> bool {
+        match *self {
+            SegwitBitcoinAddress::P2WPKH(ref mainnet, _) => *mainnet,
+            SegwitBitcoinAddress::P2WSH(ref mainnet, _) => *mainnet,
+            SegwitBitcoinAddress::P2TR(ref mainnet, _) => *mainnet,
+        }
+    }
+
+    pub fn bech32_variant(&self) -> bech32::Variant {
+        match self.witness_version() {
+            SEGWIT_V0 => bech32::Variant::Bech32,
+            _ => bech32::Variant::Bech32m,
+        }
+    }
+
+    pub fn to_bech32_hrp(&self, hrp: &str) -> String {
+        let bytes = self.to_bytes();
+        let mut bytes_u5: Vec<u5> = vec![u5::try_from_u8(self.witness_version())
+            .expect("FATAL: bad witness version does not fit into a u5")];
+        bytes_u5.extend_from_slice(&bytes.to_base32());
+        let addr = bech32::encode(&hrp, bytes_u5, self.bech32_variant())
+            .expect("FATAL: could not encode segwit address");
+        addr
+    }
+
+    pub fn to_bech32(&self) -> String {
+        let hrp = segwit_hrp(self.is_mainnet());
+        self.to_bech32_hrp(hrp)
+    }
+
+    pub fn from_bech32(s: &str) -> Option<SegwitBitcoinAddress> {
+        let (hrp, quintets, variant) = bech32::decode(s)
+            .map_err(|e| {
+                test_debug!("Failed to decode '{}': {:?}", s, &e);
+                e
+            })
+            .ok()?;
+
+        let mainnet = if hrp == SEGWIT_MAINNET_HRP {
+            Some(true)
+        } else if hrp == SEGWIT_TESTNET_HRP {
+            Some(false)
+        } else {
+            test_debug!("Unrecognized hrp '{:?}'", &hrp);
+            None
+        }?;
+
+        if quintets.len() == 0 || quintets.len() > 65 {
+            test_debug!("Invalid prog length: {}", quintets.len());
+            return None;
+        }
+
+        let version: u8 = quintets[0].into();
+        let mut prog = Vec::with_capacity(quintets.len());
+        prog.append(&mut quintets[1..].to_vec());
+
+        let bytes = Vec::from_base32(&prog)
+            .map_err(|e| {
+                test_debug!("Failed to decode quintets: {:?}", &e);
+                e
+            })
+            .ok()?;
+
+        match (variant, version, bytes.len()) {
+            (bech32::Variant::Bech32, SEGWIT_V0, 20) => {
+                let mut bytes_20 = [0u8; 20];
+                bytes_20.copy_from_slice(&bytes[0..20]);
+                Some(SegwitBitcoinAddress::P2WPKH(mainnet, bytes_20))
+            }
+            (bech32::Variant::Bech32, SEGWIT_V0, 32) => {
+                let mut bytes_32 = [0u8; 32];
+                bytes_32.copy_from_slice(&bytes[0..32]);
+                Some(SegwitBitcoinAddress::P2WSH(mainnet, bytes_32))
+            }
+            (bech32::Variant::Bech32m, SEGWIT_V1, 32) => {
```

Comment:
@jcnelson is P2WSH intended to be SEGWIT_V1 / bech32m here? In a different line above, it looks like `P2WPKH` and `P2WSH` are always `SEGWIT_V0`:
```rs
SegwitBitcoinAddress::P2WPKH(..) | SegwitBitcoinAddress::P2WSH(..) => SEGWIT_V0,
```

Reply:
NVM! I misread the expression, looks good!